velox/common/caching/tests/SsdFileTest.cpp

```diff
@@ -795,6 +798,135 @@ TEST_F(SsdFileTest, ssdReadWithoutChecksumCheck) {
+        VELOX_FAIL("Inject hook read failure");
+      }
+      if (injectDataWriteError &&
+          op->type == FaultFileOperation::Type::kWrite) {
+        VELOX_FAIL("Inject hook write failure");
+      }
+      return;
+    }
+    if (op->path.rfind(logFileName) == (op->path.size() - logFileName.size())) {
+      // Operation on eviction log file.
+      VELOX_FAIL("Inject hook read failure");
+    }
+  });
+
+  {
+    // Write a set of cache entries.
+    auto pins = makePins(
+        retainFile.id(), 0, 4096, 2048 * 1025, SsdFile::kRegionSize / 2);
+    ssdFile_->write(pins);
+
+    // With write error injected, no entry has been written to SSD cache and the
+    // error has recorded as a write ssd error.
+    SsdCacheStats statsWithWriteErrorInjected;
+    ssdFile_->updateStats(statsWithWriteErrorInjected);
+
+    EXPECT_GT(statsWithWriteErrorInjected.writeSsdErrors, 0);
+    EXPECT_EQ(statsWithWriteErrorInjected.entriesWritten, 0);
+
+    // Without write error injected, the data was cached to SSD successfully.
+    injectDataWriteError = false;
+    ssdFile_->write(pins);
+
+    SsdCacheStats statsWithoutWriteErrorInjected;
+    ssdFile_->updateStats(statsWithoutWriteErrorInjected);
+
+    EXPECT_EQ(
+        statsWithoutWriteErrorInjected.writeSsdErrors,
+        statsWithWriteErrorInjected.writeSsdErrors); // No new error occurred.
+    EXPECT_GT(statsWithoutWriteErrorInjected.entriesWritten, 0);
+
+    // Load the ssd pins by reading the ssd cache..
+    std::vector<SsdPin> ssdPins;
+    ssdPins.reserve(pins.size());
+    for (auto& pin : pins) {
+      ssdPins.push_back(ssdFile_->find(RawFileCacheKey{
+          pin.entry()->key().fileNum.id(), pin.entry()->key().offset}));
+    }
+    VELOX_ASSERT_THROW(
+        ssdFile_->load(ssdPins, pins), "Inject hook read failure");
+    injectDataReadError = false;
+    ssdFile_->load(ssdPins, pins);
+  }
+
+  // Populate ssd cache with two files.
+  std::vector<TestEntry> allEntries;
+  for (auto startOffset = SsdFile::kRegionSize;
+       startOffset <= kSsdSize / 2 - SsdFile::kRegionSize;
+       startOffset += SsdFile::kRegionSize) {
+    auto pins = makePins(
+        retainFile.id(),
+        startOffset,
+        4096,
+        2048 * 1025,
+        SsdFile::kRegionSize / 2);
+    ssdFile_->write(pins);
+    readAndCheckPins(pins);
+  }
+
+  for (auto startOffset = kSsdSize / 2;
+       startOffset <= kSsdSize - SsdFile::kRegionSize;
+       startOffset += SsdFile::kRegionSize) {
+    auto pins = makePins(
+        evictFile.id(),
+        startOffset + SsdFile::kRegionSize,
+        4096,
+        2048 * 1025,
+        SsdFile::kRegionSize / 2);
+    ssdFile_->write(pins);
+    readAndCheckPins(pins);
+  }
+
+  SsdCacheStats statsWithoutErrorInjected;
+  ssdFile_->updateStats(statsWithoutErrorInjected);
+  ASSERT_GT(statsWithoutErrorInjected.entriesWritten, 0);
+  ASSERT_GT(statsWithoutErrorInjected.entriesRead, 0);
+
+  // Remove one file from the ssd cache to trigger eviction.
```

Comment:
Can you separate the tests:
```
1 data file write injection, and we expect the failed written space is unusable until eviction?
2 data file read injection, and we expect data is readable after clearing the fault injection
3 log write injection
4 log read injection
```

Reply:
for 1: the written space will still be usable as the write operation has failed and no data is written into the space.

Reply:
I've break the test into 2 as the read/write test for the same file share a lot of common code.